src/helpers.rs

```diff
@@ -0,0 +1,76 @@
+use crate::protocol::{message::Message, payload::Version};
+
+use tokio::{
+    io,
+    net::{TcpListener, TcpStream},
+};
+
+use std::net::SocketAddr;
+
+/// Waits until an expression is true or times out.
+///
+/// Uses polling to cut down on time otherwise used by calling `sleep` in tests.
+#[macro_export]
+macro_rules! wait_until {
+    ($limit_secs: expr, $condition: expr $(, $sleep_millis: expr)?) => {
+        let now = std::time::Instant::now();
+        loop {
+            if $condition {
+                break;
+            }
+
+            // Default timout.
+            let sleep_millis = 10;
+            // Set if present in args.
+            $(let sleep_millis = $sleep_millis;)?
+            tokio::time::sleep(std::time::Duration::from_millis(sleep_millis)).await;
+            if now.elapsed() > std::time::Duration::from_secs($limit_secs) {
+                panic!("timed out!");
+            }
+        }
+    };
+}
+
+pub async fn handshake(listener: TcpListener) -> io::Result<TcpStream> {
```

Comment:
I find the naming of these guys ambiguous. Maybe `wait_for_handshake`, and `initiate_handshake`? 

I'm not quite convinced by suggestion either. Thoughts?

Reply:
I'll give it a think (this is likely not the final API), and I'll add a touch of documentation as well.